Added alternate method for generating Face3D sub-faces

michizhou opened the pull request:
Added alternate method on Face3D class to generate rectangular glazing from a base rectangular surface using a height and width, porting original Javascript code to Python. Also added tests to confirm that new Face3D class method is working properly.

Fixes #54

chriswmackey commented:
@michizhou ,
Thank you for doing this. I will try to review this as soon as I get the chance but a few quick comments are:
1. Your own test is failing because it appears to have a syntax error (https://travis-ci.org/ladybug-tools/ladybug-geometry/jobs/605222303?utm_medium=notification&utm_source=github_status)
2. I think there's a lot you can do to make the code more Pythonic including list comprehension and [ternary operators](https://en.wikipedia.org/wiki/%3F:#Python)
3. We don't use the word "widow" in the geometry library (we just refer to things as sub_faces)
4. You will probably need to make this method hidden since it is not applicable to all Face3Ds and it only works for the portion of a Face3D that is a rectangle.

michizhou commented:
@chriswmackey I just recently updated my code for the new sub-face dimensions method as per your suggestions such that it is more Pythonic and rewrote the test to deal with the build error. I also replaced all instances of "window" in my method and test with "sub_face" to accurately reflect the fact that "window" isn't in the geometry library. I haven't made the method fully hidden (with "__" before method name) since it becomes untestable then, so I just put "_" before the method name so that it's designated as partially hidden while still being testable. 

Let me know if these changes are sufficient and if not what further changes need to be made.

chriswmackey commented on ladybug_geometry/geometry3d/face.py:
```diff
@@ -1151,6 +1151,94 @@ def sub_faces_by_ratio_sub_rectangle(self, ratio, sub_rect_height, sill_height,
+        Returns:
+            A list of Face3D objects for sub faces.
+        """
+        sub_face_height_final = rect_height if sub_face_height > rect_height else sub_face_height
```
I think we will need a tolerance factor in here. The original Javascript wasn't producing geometry for EnergyPlus and so we didn't have to worry about windows going all of the way to the edge of of the parent face but this is illegal in EnergyPlus. So I think this would have to be:
```
sub_face_height_final = rect_height - 0.01 if sub_face_height > rect_height - 0.01 else sub_face_height

chriswmackey commented on ladybug_geometry/geometry3d/face.py:
```diff
@@ -1151,6 +1151,94 @@ def sub_faces_by_ratio_sub_rectangle(self, ratio, sub_rect_height, sill_height,
+            A list of Face3D objects for sub faces.
+        """
+        sub_face_height_final = rect_height if sub_face_height > rect_height else sub_face_height
+        sill_height_final = 0 if sill_height < 0.01 * rect_height else rect_height-0.1 if sill_height > rect_height-0.1 else sill_height
```
Same thing here about not going to the edge of a parent Surface. It shouldn't be possible to get a 0 sill height.

chriswmackey commented on ladybug_geometry/geometry3d/face.py:
```diff
@@ -1151,6 +1151,94 @@ def sub_faces_by_ratio_sub_rectangle(self, ratio, sub_rect_height, sill_height,
+
+        max_width_break_up = wall_length/2
+        num_divisions = round(wall_length/dist_breakup) if wall_length > dist_breakup/2 else 1
+        if sub_face_width < max_width_break_up and num_divisions * sub_face_width <= wall_length and num_divisions != 1:
```
Some in-line comments would be extremely helpful here to understand what you are going to do when these criteria are True.

chriswmackey commented on ladybug_geometry/geometry3d/face.py:
```diff
@@ -1151,6 +1151,94 @@ def sub_faces_by_ratio_sub_rectangle(self, ratio, sub_rect_height, sill_height,
+            final_glz_coords = [[line[1], line[0], [line[0][0], 0, line[0][2] + sub_face_height_final], [line[1][0], 0, line[1][2] + sub_face_height_final]]
+                                for line in sub_face_lines_start]
+            glz_area = sub_face_height_final * sub_face_width * len(sub_face_lines_start)
+        else:
```
Same thing here about in-line comments.

chriswmackey commented on ladybug_geometry/geometry3d/face.py:
```diff
@@ -1151,6 +1151,94 @@ def sub_faces_by_ratio_sub_rectangle(self, ratio, sub_rect_height, sill_height,
+            sub_face_line_scale = sub_face_width / div_dist
+
+            sub_face_lines_start = [[[x[0][0] - ((x[0][0]-y[0])*(1-sub_face_line_scale)), 0, x[0][2]],[x[1][0] + ((y[0]-x[1][0])*(1-sub_face_line_scale)), 0, x[0][2]]]
+                                    for x, y in zip(sub_face_lines_start, line_cent_pt)]
```
I would recommend using other ladybug_geometry objects here like `LineSegment3Ds`, which you can then extrude into `Face3D`s that you can return:
https://github.com/ladybug-tools/ladybug-geometry/blob/master/ladybug_geometry/geometry3d/line.py#L9

chriswmackey commented on ladybug_geometry/geometry3d/face.py:
```diff
@@ -1151,6 +1151,94 @@ def sub_faces_by_ratio_sub_rectangle(self, ratio, sub_rect_height, sill_height,
+                [sub_face_lines_start[1][0], 0, sub_face_lines_start[1][2] + sub_face_height_final]
+            ]
+        glazing_ratio = glz_area/(wall_length * rect_height)
+        return self.sub_faces_by_ratio(glazing_ratio)
```
I don't understand. You go through all of this effort to calculate the actual geometry of the sub-faces and then you just throw it away to use another method?  I thought the purpose of this method was to return a list of `Face3D` objects representing sub-faces at a given width and height. Is this just a placeholder for now?

chriswmackey commented on ladybug_geometry/geometry3d/face.py:
```diff
@@ -1151,6 +1151,94 @@ def sub_faces_by_ratio_sub_rectangle(self, ratio, sub_rect_height, sill_height,
+            rect_height: A number for the target height of the output rectangles. Note that, if the ratio is too large for the height,
+                the ratio will take precedence and the sub-rectangle height will
+                be larger than this value.
+            changed_var: string indicating the variable being changed 
```
I am at a loss to understand what this variable is or why it is important.  I feel like it might be a relic from the Javascript that doesn't really have a purpose here. Could we just remove it from the inputs and always assume the same value for this?

michizhou replied:
In that case, what should I put as the body of the first conditional (Line 1183)? Should it be the line that updates `sub_face_height_final` (Line 1185), or the one that updates `sill_height_final` (Line 1187)?

chriswmackey replied:
I can see that the code was a relic from the old javascript and it was trying to be sensitive to which input the user had just changed on the UI (so it is not relevant here). Let's give priority to the window height over the sill height so the sill height is what should be adjusted and line 1187 is the one we should use.

michizhou commented:
@chriswmackey I have just pushed the refactoring changes I made to the former Face3D alternate method in order to integrate it with the rest of the Face3D class methods. I want to take this chance to apologize for taking so long to get back to you with these changes after your last review, as well as for not being more careful with following style guidelines and coding standards more rigorously when making the changes from my last commit so as to make you have to write such an extensive code review when I clearly should have known better before pushing.

I hope these most recent changes meet the standards of the `ladybug_geometry` codebase and are sufficient for your approval. Please let me know if further changes do need to be made so that I can take care of them in a timely manner. Thank you again for your time.

michizhou commented:
@chriswmackey I just wanted to follow up with you about my most recent changes. If you could look over them whenever you are able, I would greatly appreciate it. Thank you.

michizhou commented:
@chriswmackey I know it has been over two months since I last contacted you, but I wish to follow up regarding the latest changes I made to this PR. I would greatly appreciate if you could look over them at your earliest convenience since I want to get the PR merged in a timely manner and have my contribution count toward Hacktoberfest 2019. Thank you.

claassistantio commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ladybug-tools/ladybug-geometry?pullRequest=66) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ladybug-tools/ladybug-geometry?pullRequest=66) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ladybug-tools/ladybug-geometry?pullRequest=66) it.</sub>

chriswmackey commented:
I'm very sorry that this fell off my radar, @michizhou . I will take a look as soon as I can (hopefully today). Can you resolve any of the comments that you have addressed with changes in your code?

michizhou commented:
@chriswmackey I have already tried my best to address all the comments put forward in your initial review, which are reflected in my latest pushed code changes. I will look forward to your next review in order to take care of any lingering issues in the code.

chriswmackey commented on ladybug_geometry/geometry3d/face.py:
```diff
@@ -1151,6 +1151,108 @@ def sub_faces_by_ratio_sub_rectangle(self, ratio, sub_rect_height, sill_height,
             sub_faces.extend(face.sub_faces_by_ratio(ratio))
         return sub_faces
 
+    def _sub_faces_by_dimensions(self, base_plane, rect_base, rect_height,
```
It looks like you aren't making use of `self` at all in the method here so I think you should remove it and make it a `@staticmethod` like you see here:
https://github.com/ladybug-tools/ladybug-geometry/blob/master/ladybug_geometry/geometry3d/face.py#L1304-L1305

chriswmackey commented on ladybug_geometry/geometry3d/face.py:
```diff
@@ -1151,6 +1151,108 @@ def sub_faces_by_ratio_sub_rectangle(self, ratio, sub_rect_height, sill_height,
+        # if sub_face_height > rect_height, set it to just under rect_height
+        sub_face_height_final = (
+            rect_height - 0.01 if sub_face_height > rect_height - 0.01
+            else sub_face_height)
```
I appreciate you trying to respect the 89 character limit but you don't need parentheses here and a better way to write it would be with `\` at the end of lines that are continued on the next one. [Here's an example](https://github.com/ladybug-tools/ladybug-geometry/blob/master/ladybug_geometry/geometry3d/face.py#L1366-L1367). Also, I think that we should make this like so:
```
rect_height - 0.01 * rect_height if sub_face_height > rect_height else sub_face_height
```
This method will get used in all sorts of units systems and an absolute offset of 0.01 might not be practical if the units are millimeters, for example.

chriswmackey commented on ladybug_geometry/geometry3d/face.py:
```diff
@@ -1151,6 +1151,108 @@ def sub_faces_by_ratio_sub_rectangle(self, ratio, sub_rect_height, sill_height,
+            else sub_face_height)
+        # if sill_height is close to 0, set it to just above 0
+        sill_height_final = (
+            0.01 * rect_height if sill_height < 0.01 * rect_height else sill_height)
```
Same comment here about parentheses.

chriswmackey commented on ladybug_geometry/geometry3d/face.py:
```diff
@@ -1151,6 +1151,108 @@ def sub_faces_by_ratio_sub_rectangle(self, ratio, sub_rect_height, sill_height,
+        num_divisions = (
+            round(rect_base / dist_breakup) if rect_base > dist_breakup / 2 else 1)
+        if (sub_face_width < max_width_break_up and
+            num_divisions * sub_face_width <= rect_base):
```
This line should be indented one more or you can just move it to the line above if you can get it under 89 characters.

chriswmackey replied:
Also, I would not use parentheses here.

chriswmackey commented on ladybug_geometry/geometry3d/face.py:
```diff
@@ -1151,6 +1151,108 @@ def sub_faces_by_ratio_sub_rectangle(self, ratio, sub_rect_height, sill_height,
+
+        max_width_break_up = rect_base / 2
+        num_divisions = (
+            round(rect_base / dist_breakup) if rect_base > dist_breakup / 2 else 1)
```
Same comment about parentheses

chriswmackey commented on ladybug_geometry/geometry3d/face.py:
```diff
@@ -1151,6 +1151,108 @@ def sub_faces_by_ratio_sub_rectangle(self, ratio, sub_rect_height, sill_height,
+            total_dist = remainder / 2
+
+            while total_dist < rect_base:
+                total_dist += div_dist
```
I just realized that we could get thrown into an infinite loop here if someone inputs a negative number or 0 for `dist_breakup`. Maybe we should add an `assert` statement at the top of the function just to be sure that the `dist_breakup` is always positive.

chriswmackey replied:
It looks like you addressed this one.

chriswmackey commented on ladybug_geometry/geometry3d/face.py:
```diff
@@ -1151,6 +1151,108 @@ def sub_faces_by_ratio_sub_rectangle(self, ratio, sub_rect_height, sill_height,
+            line_cent_pt = [
+                Point3D(line.p2.x + ((line.p1.x - line.p2.x) / 2), 0, line.p1.z)
+                for line in sub_face_lines_start
+            ]
```
This code will be a little more readable if you use `LineSegment3D.point_at(0.5)` in order to get the midpoint. [You can see this method here](https://github.com/ladybug-tools/ladybug-geometry/blob/master/ladybug_geometry/geometry3d/line.py#L186-L193)

chriswmackey commented on ladybug_geometry/geometry3d/face.py:
```diff
@@ -1151,6 +1151,108 @@ def sub_faces_by_ratio_sub_rectangle(self, ratio, sub_rect_height, sill_height,
+                Point3D(a.p2.x + ((b.x - a.p2.x) * (1 - sub_face_line_scale)),
+                        0, a.p1.z))
+                for a, b in zip(sub_face_lines_start, line_cent_pt)
+            )
```
This for loop with be a bit more readable if you use 'LineSegment3D.scale(sub_face_line_scale, midpoint)`. Where `midpoint` is the line segment center point that you just calculated. [You can see this method here](https://github.com/ladybug-tools/ladybug-geometry/blob/master/ladybug_geometry/geometry3d/line.py#L139-L147).

chriswmackey commented on ladybug_geometry/geometry3d/face.py:
```diff
@@ -1151,6 +1151,108 @@ def sub_faces_by_ratio_sub_rectangle(self, ratio, sub_rect_height, sill_height,
+                                  base_plane) for line in sub_face_lines_start]
+        else: # make a single sub-rectangle at an appropriate sill height
+            if sub_face_width > rect_base:
+                sub_face_width = rect_base
```
I think we need to offset the width a little here since having a sub-face that goes all of the way to the edge of the parent rectangle could break some of the workflows. So this should be:
```
if sub_face_width >= rect_base:
    sub_face_width = rect_base * 0.98
```

chriswmackey commented:
Thanks for addressing most of the comments @michizhou . Can you address the last two that are still unresolved before we merge?

michizhou commented:
@chriswmackey I thought I already did. Though the conversations may not be marked as resolved, I believe I did make the appropriate changes to address them.

michizhou commented:
@chriswmackey I thought I already did. Though the conversations may not be marked as resolved, I believe I did make the appropriate changes to address them.

chriswmackey commented:
@michizhou ,

I did not mark them as resolved because:

1. You are not making use of `LineSegment3D.point_at(0.5)` in order to get the midpoint of each line segment you generate. If it's not clear, [I would like you to use this method here](https://github.com/ladybug-tools/ladybug-geometry/blob/master/ladybug_geometry/geometry3d/line.py#L186-L193)

2. You are not making use of `LineSegment3D.scale(sub_face_line_scale, midpoint)`.  If it's not clear, [I would like you to use this method here](https://github.com/ladybug-tools/ladybug-geometry/blob/master/ladybug_geometry/geometry3d/line.py#L139-L147)

michizhou commented:
@chriswmackey My apologies for the misunderstanding. I have simplified the code accordingly. It should be ready to go now, barring any other changes I may have overlooked.

chriswmackey commented:
Merged! Welcome to the [contributors list](https://github.com/ladybug-tools/ladybug-geometry/graphs/contributors), @michizhou !
